include/net/buf.h

```diff
@@ -479,70 +492,155 @@ struct net_buf_pool {
  *  this needs to happen with the help of a separate pointer rather than an
  *  extern declaration.
+ *
+ *  The data payload of the buffers will be allocated from the heap using
+ *  k_malloc, so CONFIG_HEAP_MEM_POOL_SIZE must be set to a positive value.
+ *  This kind of pool does not support blocking on the data allocation, so
+ *  the timeout passed to net_buf_alloc will be always treated as K_NO_WAIT
+ *  when trying to allocate the data. This means that allocation failures,
+*   i.e. NULL returns, must always be handled cleanly.
  *
  *  If provided with a custom destroy callback this callback is
  *  responsible for eventually calling net_buf_destroy() to complete the
  *  process of returning the buffer to the pool.
  *
- *  @param _name     Name of the pool variable.
- *  @param _count    Number of buffers in the pool.
- *  @param _size     Maximum data size for each buffer.
- *  @param _ud_size  Amount of user data space to reserve.
- *  @param _destroy  Optional destroy callback when buffer is freed.
+ *  @param _name      Name of the pool variable.
+ *  @param _count     Number of buffers in the pool.
+ *  @param _destroy   Optional destroy callback when buffer is freed.
  */
-#define NET_BUF_POOL_DEFINE(_name, _count, _size, _ud_size, _destroy)        \
-	static struct {                                                      \
-		struct net_buf buf;                                          \
-		u8_t data[_size] __net_buf_align;	                     \
-		u8_t ud[ROUND_UP(_ud_size, 4)] __net_buf_align;              \
-	} _net_buf_##_name[_count] __noinit;                                 \
-	struct net_buf_pool _name __net_buf_align                            \
-			__in_section(_net_buf_pool, static, _name) =         \
-		NET_BUF_POOL_INITIALIZER(_name, _net_buf_##_name,            \
-					 _count, _size, _ud_size, _destroy)
+#define NET_BUF_POOL_HEAP_DEFINE(_name, _count, _destroy)                     \
+	static struct net_buf net_buf_##_name[_count] __noinit;               \
+	struct net_buf_pool _name __net_buf_align                             \
+			__in_section(_net_buf_pool, static, _name) =          \
+		NET_BUF_POOL_INITIALIZER(_name, &net_buf_heap_alloc,          \
+					 net_buf_##_name, _count, _destroy)
+
+struct net_buf_pool_fixed {
+	size_t data_size;
+	u8_t *data_pool;
+};
+
+extern const struct net_buf_data_cb net_buf_fixed_cb;
 
+/** @def NET_BUF_POOL_FIXED_DEFINE
+ *  @brief Define a new pool for buffers based on fixed-size data
+ *
+ *  Defines a net_buf_pool struct and the necessary memory storage (array of
+ *  structs) for the needed amount of buffers. After this, the buffers can be
+ *  accessed from the pool through net_buf_alloc. The pool is defined as a
+ *  static variable, so if it needs to be exported outside the current module
+ *  this needs to happen with the help of a separate pointer rather than an
+ *  extern declaration.
+ *
+ *  The data payload of the buffers will be allocated from a byte array
```

Comment:
This seems like a regression. Previously, fixed-size buf pool supported blocking on alloc. It's also the baseline pool implementation. By both reasons it should support timeouts IMHO.

Reply:
The NET_BUF_POOL_FIXED pool type creates a data array with one chunk for each net_buf in the pool. Because of this there's no need to block on the data allocation since there's always guaranteed to be a data block available. The net_buf allocation itself still uses the given timeout like before.

Reply:
> The net_buf allocation itself still uses the given timeout like before.

This is good point which needs to be emphasized - that API-level net_buf_alloc() operation actually consists of net_buf header allocation and actual data backing store allocation, both of which can block (or not block) independently. I'm not sure where's the best place to describe that though. Firsts starting point is having this point in review comments indeed.

Reply:
Another issue to consider is how strictly we want to adhere to the exact timeout that's given when both net_buf and data allocation may block. Say the user gives a 10 second timeout to net_buf_alloc() and gets a net_buf in 5 seconds. The current implementation would potentially block for another 10 seconds when waiting for the data, instead of subtracting the time waited from the original timeout.

Reply:
I've updated the PR with another patch (the second one) which now subtracts the time already spent with the net_buf allocation when proceeding to allocating the data portion.

Reply:
Hmm, github doesn't show it as the second one (even though it is in the commit history. Anyway, it's "net: buf: Fix passing too big timeout to data_alloc() "

Reply:
> Another issue to consider is how strictly we want to adhere to the exact timeout

Yeah, I thought about that too. My idea was to just document that maximum wait can be 2 * timeout specified by user, because we have bigger problems to resolve for now. Well, great that you went to resolve it in a better way, that particular patch looks clean. Thanks!